### ISSUE TITLE
JuliaFinance Registry

### ISSUE THREAD

**Issue by EricForgy** [JuliaFinance/Roadmap#7]
Hi everyone 👋 

I created an experimental [JuliaFinance registry](https://github.com/JuliaFinance/JuliaFinance).

As you may know, the new package manager allows individuals and organizations to create their own registries.

I am open to suggestions, but am personally inclined to register my packages in the `JuliaFinance` registry and NOT in the `General` registry.

Organization registries are similar to packages in that users will need to `add` them in order to use them, i.e. if someone wants to use a `JuliaFinance` package, they would - just one time only - need to add the `JuliaFinance` registry via:

```julia
pkg> registry add https://github.com/JuliaFinance/JuliaFinance.git
```

From that point, adding packages from `JuliaFinance` is done as usual with

```julia
pkg> add FinancialToolbox # or whatever
```

This is a personal choice, but one you might consider if you want to contribute a package to `JuliaFinance`. You can always have your packages registered in both `General` and `JuliaFinance` and the package manager can handle that (it is pretty cool actually).

I've been bugging the `Pkg` guys recently - including submitting issue https://github.com/JuliaLang/Pkg.jl/issues/1067 and a proof of concept PR https://github.com/JuliaLang/Pkg.jl/pull/1064  - to make a small change to `Project.toml` and hence `Manifest.toml` to let `Pkg` know which registry a package calls home so that `Pkg` can automatically `add` a registry if it is a dependency of a package registered in `General`. I still want to advocate for that, but there are more urgent things to work on and we can work around the issue by manually `add`ing registries for now.

A downside to this is that IF the guys at `Pkg` do not accept this change, then in addition to users needing to manually add the `JuliaRegistry`, they cannot submit packages to the `General` registry if they have dependencies on `JuliaRegistry`. That would be a shame (bordering on a travesty), but that is an issue for another time. I'm sure it can be resolved positively when the time comes 🙏 

Cheers

**Comment by simonbyrne** [comment]
I really don't like this idea: at the very best it is a lot busy work which adds no value. At worst, people end up with constantly broken packages.

**Comment by EricForgy** [comment]
That opinion seems a bit extreme and a little disrespectful (and a lot like the opinion of your employer). What is the basis for this opinion?

I think it adds value or I wouldn't do it. With the new Registrator.jl, managing this will be simple and with the design of Pkg jl, this already works. Nothing will be broken.

**Comment by EricForgy** [comment]
PS: "Nothing will be broken." This statement is false. Of course things will break, as always even without this issue, but I mean the "process" will not be broken. Issues will come up and issues will be resolved. That is as usual. But an upside is anyone managing their own registry has a vested interest in seeing that things work so they are likely to help find and resolve issues, i.e. more eyes / resources will be available.

**Comment by simonbyrne** [comment]
As far as I know, my employer doesn’t have an opinion about whether JuliaFinance should have its own registry. (Other employees might, but these are hardly universal, or represent any sort of policy).

I’m opposed because I don’t see there being any advantages, and lots of downsides.

**Comment by EricForgy** [comment]
The biggest advantage is that we have autonomy. I / we are more than capable of managing our own registry. So far, and maybe this could change in the future, I don't see any advantage to using General either. There is no benefit in terms of discovery. Anyone interested in Julia and Finance will surely find us.

Right now, you are right. Managing this is quite manual, but that will change with Regustrator.jl. Registrator.jl was pointed out to me by Stefan as the main piece of the puzzle outstanding before we can retire METADATA and move directly to General. Registrstor also makes it simple to manage your own registries as well.

Finally, as has bern discussed with Stefan already, Pkg already supports multiple registries. You just need to manually add them one time with "pkg> registry add MyRegistry". I've gone through the Pkg code line by line (it is impressive 🙌). I am confident things will be fine once some expected initial pains are resolved.

In summary, I see benefits with little or no downside and with Registrator and Pkg, the technical is mostly sorted.

I guess we agree to disagree.

**Comment by simonbyrne** [comment]
There are other issues than just the burden of maintaining a registry:
- it makes discoverability harder for end users
- it leads to more duplication: e.g. a developer of another package may just reimplement their own currency types just to avoid having to add depending on a different registry
  - indeed, if the decision is that packages in General are not allowed to depend on outside packages, then they will have no other choice
- packages won't be part of PkgEval runs (which are used for regression tests of releases)

Of course, you are welcome to try it and see how you go (I'm sure it will be useful to shake out Pkg bugs). My main concern was about removing existing packages from METADATA/General (https://github.com/JuliaLang/METADATA.jl/pull/21573), which could cause problems for users who are already making use of the packages.

**Comment by EricForgy** [comment]
I guess we do need to agree to disagree. That is fine 😊 

I say this will not impact discovery and you say it will impact discovery, so we are talking past each other 😅 I'd guess more than 90% of discovery happens via Google and I have no doubt someone searching "Julia Finance" will find us 😊 

> * it leads to more duplication: e.g. a developer of another package may just reimplement their own currency types just to avoid having to add depending on a different registry

This remains to be seen. I fully expect `Pkg` and `General` to support registries depending on other registries, e.g. see https://github.com/JuliaLang/Pkg.jl/issues/1072.

>    * indeed, if the decision is that packages in General are not allowed to depend on outside packages, then they will have no other choice

This would be unfortunate. Tragic even. I hope this never happens.

> * packages won't be part of PkgEval runs (which are used for regression tests of releases)

I have no doubt this could be dealt with especially if https://github.com/JuliaLang/Pkg.jl/issues/1072 is closed positively.

Anyway, like I said, this is voluntary. If you want to contribute a package to JuliaFinance and register on General, that is 100% fine. In fact, I did not request removing `DayCounts` from `General` because I wanted to know you're opinion first and now I know and it is fine 😊 

`Currencies` is a bit of a special case and I understand we should all strive to be good open-source citizens, but part of that was about thinking ahead. Having some discomfort now will make things smoother for years to come. I hope / expect good things to come from the issues raised here and spread across multiple issues.

Cheers

**Comment by StefanKarpinski** [comment]
> This would be unfortunate. Tragic even. I hope this never happens.

General is and should be self-contained. Even with the ability for registries to depend on other registries, General should not have other registries as dependencies. If it does, what is the point of having those registries be separate? Why wouldn't they just be merged into General? If General depends on JuliaFinance, then JuliaFinance will be installed on every machine by default, just like General. What advantage does having them be separate give? I listed many downsides in my discourse reply, none of which have been addressed.

This seems like all you want is to not have to deal with the minimal oversight that General has and be allowed to do whatever you want in what is effectively a portion of General. In other words, you want a system where anyone who wants to—in this case you—can unilaterally take control over an arbitrary subsection of the General registry just because they created a registry.

**Comment by EricForgy** [comment]
> If General depends on JuliaFinance, then JuliaFinance will be installed on every machine by default, just like General. What advantage does having them be separate give? 

Come on @StefanKarpinski. This is a lame straw man if I ever saw one. No one, especially me, would ever consider having every Julia user on the planet have to install `JuliaFinance` in order for `General` to support a registered package that depends on `JuliaFinance`. That is ridiculous. 

@JeffBezanson, I'm disappointed in you too for supporting such a nonsense post with a thumbs up. You should think twice before doing things like that since your opinion carries weight. Stefan's strawman post was an afront and you shouldn't support it.

[Edit: If you are anyone else want to see what I actually proposed, this issue describes it in some detail:

- [[RFC] Accommodate dependencies from other registries ](https://github.com/JuliaLang/Pkg.jl/issues/1067)

]

> I listed many downsides in my discourse reply, none of which have been addressed.

I haven't addressed any downsides that you may have listed because I never suggested this and I wouldn't support it if someone else had suggested it.

What I did suggest was a way for a registry to know if a registered dependency depends on a package in a different registry.

I think I gave this example already, but I'll repeat it...

Let's say someone writes a package `FinancialToolbox` that depends on packages registered only in `JuliaFinance`. Let's say they want to register `FinancialToolbox` in `General`. Ideally, what I'd like to see is that `General` allows this to happen and when someone tries to 

```julia
pkg> add FinancialToolbox
```

`Pkg` determines that `FinancialToolbox` depends on `JuliaFinance`, checks to see if `JuliaFinance` is already available locally. If it is available, add the package per usual. If it is not available, `Pkg` can notify the person that the package cannot be added because of the dependency and maybe even offers to add the registry for the user.

> This seems like all you want is to not have to deal with the minimal oversight that General has and be allowed to do whatever you want in what is effectively a portion of General. In other words, you want a system where anyone who wants to—in this case you—can unilaterally take control over an arbitrary subsection of the General registry just because they created a registry.

This doesn't even deserve comment. If you keep making statements like this, I will moderate accordingly.

**Comment by EricForgy** [comment]
For reference, here is a semi-detailed description of what I was proposing:

- [[RFC] Accommodate dependencies from other registries](https://github.com/JuliaLang/Pkg.jl/issues/1067)

There is nothing in there that would imply every user on the planet would need to install every registry dependency. That is like saying every user would need to install every package in `General`. Come on 🤷‍♂️

**Comment by simonbyrne** [comment]
> Pkg determines that FinancialToolbox depends on JuliaFinance, checks to see if JuliaFinance is already available locally. If it is available, add the package per usual. If it is not available, Pkg can notify the person that the package cannot be added because of the dependency and maybe even offers to add the registry for the user.

From the point of the end user, that seems strictly worse than having everything in General. 

There are also serious security concerns as it makes it easier for malicious packages to propagate.

**Comment by EricForgy** [comment]
> There are also serious security concerns as it makes it easier for malicious packages to propagate.

This assumes the curators of the registry are less qualified to manage their own packages than those of `General`, which would be a stretch.

**Comment by Keno** [comment]
Hi @EricForgy, perhaps you could clarify your intent here. The only thing that's special about the `General` registry is that it is the registry that happens to ship with the Julia distribution from julialang.org and is subject to the editorial control of the core julia team (however minimal that may be at this point - though experience with other package ecosystems shows that more may be required in the future). People seem to have gotten the impression that the goal was to bypass these editorial control by making packages available through General without them having to be registered there. If that is the goal, you might see why people would be unsympathetic. I do think there is something to be said about spreading out the responsibility for curating `General`. Registries that get wholesale imported into `General` might be one way to do it (under the assumptions that the same editorial controls would apply, similar to the Linux maintainership model), but there are others (e.g. permission bits on Registrator once that comes online). If the second is the goal, I think perhaps resuming this discussion in a few weeks once the transition to `General` is more complete may be prudent.

**Comment by EricForgy** [comment]
Thanks @Keno. 

> Hi @EricForgy, perhaps you could clarify your intent here.

Sure. I will try.

> People seem to have gotten the impression that the goal was to bypass these editorial control by making packages available through General without them having to be registered there.

I don't know who those people might be. I am certainly not one of them.

> If that is the goal, you might see why people would be unsympathetic.

Sure. Absolutely.

> I do think there is something to be said about spreading out the responsibility for curating `General`. Registries that get wholesale imported into `General` might be one way to do it (under the assumptions that the same editorial controls would apply, similar to the Linux maintainership model), but there are others (e.g. permission bits on Registrator once that comes online).

I agree. That would make sense, but is a little orthogonal to what is my intent.

> If the second is the goal, I think perhaps resuming this discussion in a few weeks once the transition to `General` is more complete may be prudent.

There is a third case. An organization just wants to maintain a registry of its own packages. This is my intent.

It is not my intention to register a package on `General` that depends on `JuliaFinance` in order to bypass any controls. That defeats the purpose of having `JuliaFinance`. I do like the fact that my packages in `JuliaFinance` can depend on `General`. I said it before and I'll keep saying it, I've spent some decent time looking through the guts of Pkg.jl and I have extreme high regards for it. The fact this is even possible to talk about is a testament to that.

The point of https://github.com/JuliaLang/Pkg.jl/issues/1067 was to support others who might like to register a finance-related package in `General` that has a dependency on `JuliaFinance`. This is a policy issue and is the responsibility of those in charge of `General` to work out. I can voice my opinion, but I'll live with whatever is decided.

In any case, I already said that this discussion can wait until METADATA is retired here: https://github.com/JuliaLang/METADATA.jl/pull/21573#issuecomment-466753585

In the meantime, KristofferC and I have come to an agreement on how to proceed with the specific case of Currencies.jl here: https://github.com/JuliaLang/METADATA.jl/pull/21715#issuecomment-467098664

Cheers

**Comment by StefanKarpinski** [comment]
There's a significant technical matter that you seem to be missing here, which I tried to explain in the discourse thread, but which does not seem to have been sufficiently clear. You talk about this as if you can know before installing a separate registry whether you need to install a package from that registry or not, but that is not the case:

* You need a complete graph of possible package versions in order to decide what versions of packages are going to be installed.
* In order to get a complete graph of possible versions, in your proposed scheme, one needs to install other registries.
* Therefore, to even decide if one _might_ install packages from other registries, one needs to install those registries.
* Effectively, all registries that are even potentially reachable from a given package must be installed.

A very plausible scenario in your proposed design is that a user asks to install a package in General and is prompted to install multiple external registries regardless of whether they end up actually using anything from those registries or not. Suppose, for example, that an older version of some package used to take a "kitchen sink" approach and have all of the dependencies (DataFrames and Gadfly come to mind). Even if later versions of those packages slim down and have very few dependencies, if _any_ of the old dependencies have any versions (compatible or not) that depend on any external registries, then you need to install all of those registries when considering installing any version of one of these packages. This does not seem like a good user experience. The option of automatically installing external registries for them that they did not request or approve and may not trust seems worse and makes those external registries effectively part of General.

> There is a third case. An organization just wants to maintain a registry of its own packages. This is my intent.

But why? What is the benefit of JuliaFinance having its own registry _and_ having packages in General depend on those packages. Why do you want this instead of just registering JuliaFinance packages in General?

**Comment by EricForgy** [comment]
> You talk about this as if you can know before installing a separate registry whether you need to install a package from that registry or not, but that is not the case:

I explained how this could be done in my PR. 

https://github.com/JuliaLang/Pkg.jl/issues/1067

You don't need to guess. You don't need a tree. With a minor change to Pkg.jl, maintainers can just tell you.

**Comment by EricForgy** [comment]
PS: Thanks to some super help from KristofferC, Currencies.jl has been restored 🙏

**Comment by StefanKarpinski** [comment]
No, that PR does not address this as far as I can tell. How does it avoid needing to fetch other registries even if they will not end up being used?

You also have still not given any reason why splitting the dependency graph into registries that have mutual dependencies between them is beneficial. Why does that make life better for Julia users?

**Comment by EricForgy** [comment]
> No, that issue does not address this as far as I can tell. How does it prevent needing to fetch other registries even if they will not end up being used?

The same way you don't need to fetch packages that aren't used. 

> You also have still not given any reason why splitting the dependency graph into registries that have mutual dependencies between them is beneficial. Why does that make life better for Julia users?

Sorry, but I don't really care all that much. Do it if you want to do it. If you don't want to do it, don't do it. Either way is fine for me. It is your choice. Just like it is my choice to have my own registry.

**Comment by StefanKarpinski** [comment]
> The same way you don't need to fetch packages that aren't used.

I'm afraid that does not work. The way you can decide what needs to be installed without fetching packages is because all of the information you need to do version resolution is, by design, in the registry—that's its whole purpose. If _any version_ of a package that you're considering installing some version of has a dependency in another registry, then you need that registry in order to have enough information to decide what versions _of anything_ to install. This is because registries are, by definition, where the information required to decide what to install lives. So it's very possible (and imo quite likely in practice) that you might have to install a bunch of external registries to get enough information to decide that you don't want to install anything form any of them.

**Comment by simonbyrne** [comment]
I've been in meetings most of the day, but I see this conversation has gotten rather intense.

My initial issue was the removal of existing packages from METADATA/General which seemed likely to cause problems. I'm glad that was resolved.

My comment about security was a bit brief, but my concern is that if users are encouraged to use multiple public registries, they are implicitly extending their trust to maintainers of those registries. I'm not saying our current approach to security is bullet proof, but a malicious actor having control of a registry presents a much larger threat than them having control of a package repository. There is also the risk of "left pad"-style deletion problems: for a package, as long as there is a mirror, we can just change the URL and users won't notice the difference; someone deleting or moving a registry will cause much larger problems.

**Comment by c42f** [comment]
> my concern is that if users are encouraged to use multiple public registries, they are implicitly extending their trust to maintainers of those registries

Agreed; this is a real problem present in other systems. For example, the Ubuntu PPA system is (ab)used to distribute certain important software that Canonical is unable to add to the main package archive (eg the Oracle JDK, I believe due to licensing). This is a problem if users become habituated in adding PPAs because they can easily add something malicious. The PPA system provides no help at all in representing or verifying the underlying network of trust between maintainers. This is not a good user experience.

IMO the main question to focus on here remains: How will multiple registries benefit the *users* of the Julia ecosystem? From a developer perspective, I find that registering a new version of one of my packages is by far the easiest step compared to the hard work of writing and maintaining the package!

**Comment by EricForgy** [comment]
For sure, we could make mistakes, but I have faith in this community to learn from other's successes and mistakes and come up with a reasonable solution. If that means having one centralized registry, so be it. I just think that would be unfortunate, but I've already expressed my opinion. The issue is closed 😊
